samples: net: zperf: allow networking code to be relocated to RAM

danieldegrasse opened the pull request:
Enable relocation of key networking stack functions to RAM when running
the zperf example. This will enable better performance on platforms with
fast code RAM.

Some performance numbers using this PR:

Board  | Zephyr Revision                           | Code location   | TCP RX   | TCP TX   | UDP RX     | UDP TX
-------|-------------------------------------------|-----------------|----------|----------|------------|-----
RT1170 | 25c6553edde5faab3565165c27d9c1da94a081f9  | QuadSPI         | 18.4Mbps	| 12.8Mbps | 38.59Mbps	| 24.2Mbps
RT1170 | feature/zperf_relocate	                   | ITCM/QuadSPI	 | 94.0Mbps | 69.4Mbps | 89.03Mbps  | 76.7Mbps
RT1060 | 25c6553edde5faab3565165c27d9c1da94a081f9  | QuadSPI	     | 34.5Mbps | 29.0Mbps | 80.88Mbps  | 65.6Mbps
RT1060 | feature/zperf_relocate	                   | ITCM/QuadSPI	 | 93.8Mbps | 70.6Mbps | 89.03Mbps  | 73.2Mbps
RT1050 | 25c6553edde5faab3565165c27d9c1da94a081f9  | HyperFlash	     | 57.4Mbps | 55.9Mbps | 89.03Mbps  | 73.1Mbps
RT1050 | feature/zperf_relocate	                   | ITCM/Hyperflash | 94.0Mbps | 71.1Mbps | 89.03Mbps  | 73.0Mbps






Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: net: zperf: allow networking code to be relocated to RAM

Enable relocation of key networking stack functions to RAM when running
the zperf example. This will enable better performance on platforms with
fast code RAM.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
samples: net: zperf: relocate stack to RAM for iMX RT boards

Relocate network stack to RAM for iMX RT boards when running the zperf
sample. Relocating the network stack to ITCM greatly improves
performance on these platforms.

Also, remove overlays for the RT1050 and RT1060 setting the system timer
to systick. This is no longer required as this is the default timer for
these boards.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
The other alternative to this I have considered is something similar to #50792, where the network relocation options are integrated into the subsystem. My reasoning for keeping the relocation directives within the zperf sample is that a user tuning their application for performance will likely want to relocate a specific subset of the networking stack, and I feel that the decision of which portions to relocate is generally an application one.

One unfortunate side effect of this is that the NXP HAL now uses an application specific Kconfig to determine if the MCUX Ethernet driver should be relocated to RAM, which is required in order to realize the performance numbers within this PR.

rlubos reviewed (dismissed):
I'm not an expert on code relocation, but changes in the sample look ok.

ssharks commented on west.yml:
```diff
@@ -183,7 +183,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: e0116b8eae46c2d90f2c1b23baef73bb082d00bd
+      revision: pull/193/head
```
This does not look like an upstreamable change, could you use a hash here?

danieldegrasse replied:
It is not upstreamable- the relevant PR needs to be merged first

The reason this change is currently stalled is that we need to relocate the NXP HAL driver to RAM as well as the in tree ethernet driver code, and we do not want to make the relocation directive in the HAL dependent on this application defined KConfig.

ssharks reviewed (commented):
I have never worked with code relocation, but it looks like a valuable addition. It would be a said if it would be lost.
Could you look at the reference to the NXP hal, after a rebase we should be good to proceed.

carlescufi commented:
@danieldegrasse if you'd like this into 3.5 then please let me know on Discord so we can merge the HAL PR

danieldegrasse commented:
Updated this PR onto latest main, and removed the HAL changes. The HAL change does enable the Ethernet driver code to be relocated to ITCM, which improves performance further, so here are new performance numbers without this change:

Board  | Zephyr Revision                           | Code location   | TCP RX   | TCP TX   | UDP RX     | UDP TX
-------|-------------------------------------------|-----------------|----------|----------|------------|-----
RT1170 | feature/zperf_relocate	                   | ITCM/QuadSPI	 | 56Mbps | 34.26Mbps | 78.82Mbps  | 76.7Mbps
RT1060 | feature/zperf_relocate	                   | ITCM/QuadSPI	 | 72.1Mbps | 55.9Mbps | 88.78Mbps  | 75.1Mbps
RT1050 | feature/zperf_relocate	                   | ITCM/Hyperflash | 66.0Mbps | 44.1Mbps | 88.91Mbps  | 75.0 Mbps

jukkar commented on samples/net/zperf/Kconfig:
```diff
@@ -0,0 +1,22 @@
+
+source "Kconfig.zephyr"
+
+config NET_CODE_RELOCATE
```
I would prefer that the networking sample own Kconfig symbols are named so that there is no possibility to mix them with the core stack symbols. This is followed in other samples too.
So prefer:
```
config NET_SAMPLE_CODE_RELOCATE
```
and
```
config NET_SAMPLE_CODE_RAM_NAME
```

danieldegrasse replied:
Thanks for the suggestion- updated to use these names
